### PR TITLE
Development

### PR DESCRIPTION
**Pull request by amirsarebani81**

### CODE REVIEW

File: unique_ptr.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        return *this;
+    }
+
+    operator bool() { return pointer != nullptr; }
+
+    T &operator*() { return *pointer; }
+
+    T *operator->() { return pointer; }
+
+    ~unique_ptr() { pointer = nullptr; }
```
Comment (IYP-Programer-Yeah) [inline unique_ptr.hpp:44]: Search about RAII and double check what responsibilities `unique_ptr` holds.

File: unique_ptr_test.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+// 1
+TEST(UniquePtrTest, DefaultConstructor) {
+    unique_ptr<int> u_ptr;
+    ASSERT_EQ(u_ptr.get(), nullptr);
+}
+
+// 2
+TEST(UniquePtrTest, Constructor) {
+    int i = 10;
+    unique_ptr<int> u_ptr(&i);
```
Comment (IYP-Programer-Yeah) [inline unique_ptr_test.cpp:13]: Try this code with `std::unique_ptr` and see if it works. Learn about address sanitizer and try to debug the `std::unique_ptr` version of this code with address sanitizer.

File: unique_ptr_test.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    ASSERT_EQ(u_ptr.get(), &i);
+}
+
+// 3
+TEST(UniquePtrTest, MoveCostructor) {
+    int i = 10;
+    unique_ptr<int> u_ptr1(&i);
```
Comment (rasaoskuei) [inline unique_ptr_test.cpp:20]: In this unit test (also in others) what problem has unique_ptr solved?

File: unique_ptr.hpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    void swap(unique_ptr &u_ptr) { std::swap(pointer, u_ptr.pointer); }
+
+    T *get() const { return pointer; }
+
+    operator bool() const { return get() != nullptr; }
+
+    unique_ptr &operator=(const unique_ptr &u_ptr) = delete;
+
+    unique_ptr &operator=(unique_ptr &&u_ptr) {
```
Comment (IYP-Programer-Yeah) [inline unique_ptr.hpp:44]: this is not tested

**Pull request merged**